simple_certmanager/admin.py

```diff
@@ -92,6 +94,51 @@ class SigningRequestAdmin(admin.ModelAdmin):
         ),
     )
 
+    def get_urls(self):
+        urls = super().get_urls()
+        my_urls = [
+            path(
+                "dowload_csr/<int:pk>",
+                self.admin_site.admin_view(self.download_csr),
+                name="download_csr",
+            )
+        ]
+        return my_urls + urls
+
+    def download_csr(self, request, pk):
+        # Here I don't use the change permission because it won't be given after
+        # the CSR is signed and the certificate is uploaded.
+        if self.has_add_permission(request):
+            csr_instance = get_object_or_404(SigningRequest, pk=pk)
+            csr_content = csr_instance.csr
+            csr_filename = (
+                f"{slugify(csr_instance.common_name)}_{csr_instance.pk}_csr.pem"
+            )
+            return FileResponse(
+                BytesIO(csr_content.encode("ascii")),
+                as_attachment=True,
+                filename=csr_filename,
+            )
+        else:
+            messages.error(
+                request, _("You do not have permission to download this CSR.")
+            )
+            return HttpResponseRedirect(
+                reverse("admin:simple_certmanager_signingrequest_changelist")
+            )
+
+    def csr_link(self, obj: SigningRequest):
+        # Here I check that the common name is not empty
+        # because the CSR can't be generated without it and will only be generated
+        # after the object is saved the first time.
```

Comment:
Nice comment. Should we do a similar check in the download_csr view? Do I get a 404 now if there's no common_name? Is it even possible? `SigningRequest.common_name` doesn't have `blank=True`, so `""` is not a valid value.

Reply:
if there is no common_name that means we are creating the instance. The download_csr is only used after the model has been saved at least once. A saved model will hold a csr.

You won't get 404 as the link is not visible as long as there is no common_name.

![image](https://github.com/user-attachments/assets/a1fcfff5-3a8b-468b-afed-36078e45ab40)